### PR TITLE
Introducing a TimeSeries.convolve() method based on the overlap-save algorithm

### LINKED ISSUE #451: Applying FIR filters to long timeseries
Scipy seems to lack a method for applying FIR filters to very long time series (meaning longer than can be efficiently FFTed). One standard way to do this is the [overlap-save method](https://en.wikipedia.org/wiki/Overlap-save_method). It goes through the data by chunks, overlapping enough to avoid the corruption from filter wraparound.

I've implemented this using standard numpy and scipy calls. The sample code is below. I think this should be made a method of the TimeSeries class, maybe called `.convolve`, since it's equivalent to `scipy.signal.convolve`.

If this is OK, I can prepare a patch to `TimeSeries` in a more gwpy style. This could be used to close #358.

```python
def overlap_save(in1, in2, fftlen=16384):
    """Convolve a very long data series with a short one
        using the overlap-save method. Result should be
        nearly identical to scipy.signal.convolve with
        mode='same'.
        
        Inputs:
        in1 (numpy array): long data series
        in2 (numpy array): short data series (e.g. FIR filter)
        fftlen (int, optional): number of points to use for FFT
        
        Output:
        Numpy array with same length as in1"""
    # If data is short enough, just do a single convolve
    if len(in1) <= 2*fftlen:
        return sig.convolve(in1, in2, mode='same')
    
    # Find how much is corrupted by filter wraparound
    pad = int(np.ceil(len(in2)/2.))
    step = int(fftlen) - 2*pad
    # Create room for the result
    result=np.zeros(len(in1), dtype=in1.dtype)
    # Do the first chunk on its own
    result[:fftlen-pad]=sig.fftconvolve(in1[:fftlen-pad], in2, mode='same')
    # Walk through chunk by chunk
    for idx in xrange(2*step, len(in1), step):
        tmp = sig.fftconvolve(in1[idx-fftlen:idx], in2, mode='same')
        result[idx-fftlen+pad:idx-pad] = tmp[pad:-pad]
    # Clean up the last piece which may be anything smaller than step
    # It's important that we made sure to have at least one step in
    # the loop above, otherwise idx is undefined
    len_final = len(in1) - idx
    tmp = sig.fftconvolve(in1[-fftlen:], in2, mode='same')
    result[-len_final-pad:] = tmp[-len_final-pad:]
    
    return result
```

### PR DESCRIPTION
**Pull request by alurban**
@duncanmmacleod, this PR introduces a `TimeSeries.convolve()` method which efficiently applies an FIR filter to a long timeseries using the [overlap-save](https://en.wikipedia.org/wiki/Overlap–save_method) algorithm. It also adds a unit test for `TimeSeries.convolve()` which prepares a timeseries and then checks that the input data are unchanged by convolving with a delta function.

@andrew-lundgren, this fixes #451.

### CODE REVIEW

**Comment by duncanmmacleod** [comment]
@alurban, can you take a look at the failing tests? it looks like this presumes some minimal version of scipy that is beyond what the LDG clusters can support. I do plan to move away from supporting those super-old versions of numpy and scipy, but would like to at least hit 1.0.0 with LDG compatibility.

**Comment by alurban** [comment]
@duncanmmacleod, the basic issue was that prior to scipy 0.19.0, `scipy.signal.convolve()` only ever used direct convolution. I've attempted to address this by switching to `scipy.signal.fftconvolve`, which (1) still exists in subsequent scipy versions, (2) always convolves in the frequency domain, and (3) is faster than direct convolution for filters longer than ~500 taps. (In scipy >= 0.19.0, calls to `signal.convolve()` automatically invoke `fftconvolve` when the filter is long enough.)

I've also tried to address concerns from coveralls and codeclimate by cleaning up the use cases and docstrings of new functions in `gwpy.signal.filter_design`.

File: gwpy/signal/filter_design.py [hunk 1]
```diff
@@ -182,7 +183,7 @@ def truncate_transfer(transfer, ncorner=None):
     return out
 
 
-def truncate_impulse(impulse, ntaps, window='hanning'):
+def truncate_impulse(impulse, ntaps, window='hamming'):
```
Comment (duncanmmacleod) [inline gwpy/signal/filter_design.py]: what's the motivation for this change?

Comment (alurban) [reply gwpy/signal/filter_design.py]: I'm actually not sure why that's there, I can change it back if you like? (A Hanning window was the default for the old `TimeSeries.whiten` method.)

Comment (duncanmmacleod) [reply gwpy/signal/filter_design.py]: @alurban, yes please, lets keep this PR as clean as possible

Comment (alurban) [reply gwpy/signal/filter_design.py]: @duncanmmacleod, ok, I'll make that change shortly

Comment (alurban) [reply gwpy/signal/filter_design.py]: This is done with the latest commit (which I've squashed with the main commit)

**Comment by alurban** [comment]
@duncanmmacleod, I've noticed that the new whitening filter corrupts the left and right boundaries of the input signal with huge wings that go up to enormous values. I was cropping these off, but I've just pushed a change to `TimeSeries.convolve()` that instead smoothly tapers off the left and right boundaries of the input before FFT'ing, then returns a `TimeSeries` the same length as the input (and with the same `t0`). So, the whitened timeseries smoothly rises up from 0 at both ends, and the result is not cropped. This better reproduces the functionality of the original `TimeSeries.whiten()`.

In a moment I'll update the unit tests to reflect this.

**Comment by alurban** [comment]
@duncanmmacleod, the offending Travis CI build failure looks like a transient SSL timeout, probably unrelated to recent changes. Can it be restarted?

**Comment by alurban** [comment]
@duncanmmacleod, I believe I've worked out all the gotchas and this PR is ready for proper review. Below are plots of (1) the raw and whitened timeseries from Hanford for 64 seconds surrounding GW150914; (2) the whitened timeseries, zoomed in and bandpassed to show the BBH signal; and (3) the Q-transform spectrogram showing the BBH trace. All three plots were made with whitening defaults under the new changes. Note that in the spectrogram, the low-frequency power of the signal is restored, better reproducing the published spectrogram for GW150914.

150914_compare
150914_whitened_timeseries
150914_strain

**Comment by alurban** [comment]
@duncanmmacleod, per your suggestion in a private slack message, I've posted an issue (#890) and a separate pull request (#891) for all the changes from this one that do not explicitly introduce significant new functionality. That PR can be merged for a 0.12.1 point release, while this one should be saved for 0.13.0/1.0.0 since `TimeSeries.convolve()` certainly is a significant new feature.

**Comment by alurban** [comment]
@duncanmmacleod, after you merged #891 I rebased this branch against `upstream/develop` and resolved a few (minor and expected) conflicts. After pushing those changes, all the build checks have passed and this PR is ready whenever you are.

**Comment by alurban** [comment]
@duncanmmacleod, I've rebased this branch against `upstream/develop` to get all the functionality introduced in gwpy-0.12.2.

**Comment by alurban** [comment]
@duncanmacleod, yes, please merge at your discretion and I’ll rebase #905 so that it’s ready as well.
